File: python/008. String to Integer (atoi).py
class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        length = len(str)
        idx = 0
        while idx < length and str[idx] == ' ':
            idx += 1
        sign = 1
        if idx < length and (str[idx] == '+' or str[idx] == '-'):
            sign = 1 if str[idx] == '+' else -1
            idx += 1
        if not (idx < length and str[idx] >= '0' and str[idx] <= '9'): return 0
        num = 0
        while idx < length and str[idx] >= '0' and str[idx] <= '9':
            num = 10 * num + ord(str[idx]) - ord('0')
            idx += 1
        num *= sign
        num = num if num <= 2147483647 else 2147483647
        num = num if num >= -2147483648 else -2147483648
        return num

File: C++/008. String to Integer (atoi).cpp
class Solution {
public:
    int myAtoi(string str) {
        int size = str.length();
        int result=0;
        long long temp=0;
        int f=1;
        if(size==0)
        {
            return 0;
        }
        int i=0;
        while(i<size)
        {
            if(str[i]==' ')
            {
                ++i;
            }
            else if ((str[i]=='+' || str[i]=='-') && (str[i+1]<='9' && str[i+1]>='0'))
            {
                f=(str[i]=='-')? -1:1;
                for(int j=i+1;j<=i+12;j++)
                {
                    if(str[j]<='9' && str[j]>='0')
                    {
                        temp=10*temp+str[j]-'0';
                    }
                    else
                    {
                        break;
                    }
                }
                break;
            }
            else if (str[i]<='9' && str[i]>='0')
            {
                for(int j=i;j<=i+12;j++)
                {
                    if(str[j]<='9' && str[j]>='0')
                    {
                        temp=10*temp+str[j]-'0';
                    }
                    else
                    {
                        break;
                    }
                }
                break;
            }
            else
            {
                return 0;;
            }
        }
        temp*=f;
        if(temp>INT_MAX)
        {
            result=INT_MAX;
        }
        else if(temp<INT_MIN)
        {
            result=INT_MIN;
        }
        else
        {
            result = (int)temp;
        }
        return result;
    }
};
